### PR TITLE
add support for specified proxy in source GitRepository

### LINKED ISSUE #575: Flaky test TestImageUpdateAutomationReconciler_deleteBeforeFinalizer
The test flakes with the following error message:

```
--- FAIL: TestImageUpdateAutomationReconciler_deleteBeforeFinalizer (2.05s)
    update_test.go:162: 
        Unexpected error:
            <*errors.StatusError | 0xc00010a280>: 
            Operation cannot be fulfilled on imageupdateautomations.image.toolkit.fluxcd.io "test-imageupdate": the object has been modified; please apply your changes to the latest version and try again
            {
                ErrStatus: {
                    TypeMeta: {Kind: "", APIVersion: ""},
                    ListMeta: {
                        SelfLink: "",
                        ResourceVersion: "",
                        Continue: "",
                        RemainingItemCount: nil,
                    },
                    Status: "Failure",
                    Message: "Operation cannot be fulfilled on imageupdateautomations.image.toolkit.fluxcd.io \"test-imageupdate\": the object has been modified; please apply your changes to the latest version and try again",
                    Reason: "Conflict",
                    Details: {
                        Name: "test-imageupdate",
                        Group: "image.toolkit.fluxcd.io",
                        Kind: "imageupdateautomations",
                        UID: "",
                        Causes: nil,
                        RetryAfterSeconds: 0,
                    },
                    Code: 409,
                },
            }
        occurred
```

Example logs of a failed workflow run: [logs_3920.zip](https://github.com/fluxcd/image-automation-controller/files/12406313/logs_3920.zip)

### PR DESCRIPTION
**Pull request by aryan9600**
Add support for using the proxy specified in the source GitRepository using its `.spec.proxySecretRef`.

also closes #575

### CODE REVIEW

**Review by makkes — COMMENTED** [review]
Please add tests for the new behaviour.

**Comment by aryan9600** [comment]
@makkes the tests for proxy exist in `pkg/git/gogit`: https://github.com/fluxcd/pkg/blob/f775906949d5af477edc3a295b2b2283b918dca6/git/gogit/clone_test.go#L1131

**Comment by hiddeco** [comment]
@aryan9600 but that doesn't test the wiring of Secret data to actually performing the operation, which should have coverage here.

**Comment by makkes** [comment]
> @makkes the tests for proxy exist in `pkg/git/gogit`: https://github.com/fluxcd/pkg/blob/f775906949d5af477edc3a295b2b2283b918dca6/git/gogit/clone_test.go#L1131

That's great but this change adds a lot of code around the functionality from pkg/git/gogit that we should test.

**Comment by aryan9600** [comment]
@makkes @hiddeco added tests

File: internal/controller/update_test.go [hunk 1]
```diff
@@ -158,8 +160,13 @@ func TestImageUpdateAutomationReconciler_deleteBeforeFinalizer(t *testing.T) {
...
 	// NOTE: Only a real API server responds with an error in this scenario.
-	_, err := r.Reconcile(ctx, ctrl.Request{NamespacedName: client.ObjectKeyFromObject(imageUpdate)})
-	g.Expect(err).NotTo(HaveOccurred())
+	g.Eventually(func() bool {
+		_, err := r.Reconcile(ctx, ctrl.Request{NamespacedName: client.ObjectKeyFromObject(imageUpdate)})
+		if err != nil {
+			return false
+		}
+		return true
+	}, timeout).Should(BeTrue())
```
Comment (makkes) [inline internal/controller/update_test.go:169]:
```suggestion
	g.Eventually(func() error {
		_, err := r.Reconcile(ctx, ctrl.Request{NamespacedName: client.ObjectKeyFromObject(imageUpdate)})
		return err
	}, timeout).Should(Succeed())
```

This way the failure message actually tells us what failed.

**Review by makkes — APPROVED** [review]

File: Makefile [hunk 2]
```diff
@@ -23,7 +23,8 @@ CACHE := cache
 
 # Version of the source-controller from which to get the GitRepository CRD.
 # Pulls source-controller/api's version set in go.mod.
-SOURCE_VER ?= $(shell go list -m github.com/fluxcd/source-controller/api | awk '{print $$2}')
+# SOURCE_VER ?= $(shell go list -m github.com/fluxcd/source-controller/api | awk '{print $$2}')
+SOURCE_VER = 944f4cfa10c5
```
Comment (makkes) [inline Makefile:27]: Wait, this and the changes in go.mod should be reverted before merging, right?

**Review by makkes — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (aryan9600) [reply Makefile:27]: yes, this PR depends on a new release of `fluxcd/pkg/git` and SC

**Review by hiddeco — APPROVED** [review]

**Pull request merged**